### PR TITLE
[consensus] better protect sync condition

### PR DESCRIPTION
**Pull request by zekun000**
This commit does a few things to avoid that sync_to races with commit in state computer.

### CODE REVIEW

File: consensus/src/state_computer.rs [hunk 1]
```diff
@@ -201,25 +201,32 @@ impl StateComputer for ExecutionProxy {
...
+        let mut committed_round = self.write_mutex.lock().await;
+        let target_round = target.ledger_info().round();
 
         // Before the state synchronization, we have to call finish() to free the in-memory SMT
         // held by BlockExecutor to prevent memory leak.
         self.executor.finish();
 
+        // The pipeline phase already committed beyond the target synced round, just return.
+        if *committed_round >= target_round {
+            return Ok(());
```
Comment (bchocho) [inline consensus/src/state_computer.rs:219]: Do we want a counter or log here, to track if this is happening too often?

**Review by bchocho — COMMENTED** [review]
> increase the threshold of fallback to state sync while block exists locally

This isn't implemented, right? It doesn't seem necessary IMO.

**Comment by zekun000** [comment]
> This isn't implemented, right? It doesn't seem necessary IMO.

oh yeah I did that initially and decided to revert

File: consensus/src/state_computer.rs [hunk 2]
```diff
@@ -193,33 +191,37 @@ impl StateComputer for ExecutionProxy {
...
-        let _guard = self.write_mutex.lock().await;
+        let mut committed_round = self.write_mutex.lock().await;
+        let target_round = target.ledger_info().round();
 
         // Before the state synchronization, we have to call finish() to free the in-memory SMT
         // held by BlockExecutor to prevent memory leak.
         self.executor.finish();
 
+        // The pipeline phase already committed beyond the target synced round, just return.
+        if *committed_round >= target_round {
```
Comment (bchocho) [inline consensus/src/state_computer.rs:209]: Do we have to check epoch here?

File: consensus/src/state_computer.rs [hunk 3]
```diff
@@ -193,35 +191,37 @@ impl StateComputer for ExecutionProxy {
...
+        let logical_time =
+            LogicalTime::new(target.ledger_info().epoch(), target.ledger_info().round());
 
         // Before the state synchronization, we have to call finish() to free the in-memory SMT
         // held by BlockExecutor to prevent memory leak.
         self.executor.finish();
 
+        // The pipeline phase already committed beyond the target synced round, just return.
+        if *latest_logical_time >= logical_time {
+            warn!(
```
Comment (bchocho) [inline consensus/src/state_computer.rs:211]: This is a bit of a nit. I thought we established that `==` is possible even without this race. It makes sense to return right away in this case (nothing more to do), but maybe only warn in the `>` case?

**Review by bchocho — APPROVED** [review]
LGTM. One more comment below

File: consensus/src/state_computer.rs (hunk 3 above)
Comment (zekun000) [reply consensus/src/state_computer.rs:211]: it's similarly unexpected, just state sync would not return error previously. it's expected we only sync to future round

File: consensus/src/state_computer.rs [hunk 4]
```diff
@@ -193,35 +191,37 @@ impl StateComputer for ExecutionProxy {
...
+        let mut latest_logical_time = self.write_mutex.lock().await;
+        let logical_time =
+            LogicalTime::new(target.ledger_info().epoch(), target.ledger_info().round());
 
         // Before the state synchronization, we have to call finish() to free the in-memory SMT
         // held by BlockExecutor to prevent memory leak.
         self.executor.finish();
 
+        // The pipeline phase already committed beyond the target synced round, just return.
+        if *latest_logical_time >= logical_time {
```
Comment (sitalkedia) [inline consensus/src/state_computer.rs:210]: Can you explain why is this needed? Even if the committed version is greater than the target, that shouldn't cause any issue for state sync?

**Review by sitalkedia — COMMENTED** [review]
Can you add more context about the issue to the summary?

File: consensus/src/state_computer.rs (hunk 4 above)
Comment (zekun000) [reply consensus/src/state_computer.rs:210]: state sync would return error and cause the buffer manager to not get reconstructed, but the problem is worse with assertion in quorum store that panics https://github.com/aptos-labs/aptos-core/blob/5af60240f125b2596ebd93ceb52fad29f284bbd9/consensus/src/quorum_store/batch_reader.rs#L342

**Review by sitalkedia — APPROVED** [review]
LGTM, please update the summary accordingly before landing.
